keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,389 @@
+for KUDO will make selecting of resources much easier, rather than having to
+use magical strings with templates. Future iterations of this will reduce the
+complexity of selecting resources to run commands and files on.
+
+Steps in a bundle run serially. To prevent the KUDO controller manager from
+crashing, the collector process runs in another pod as a job. **TODO**:
+Bundle collection, CRD, do we take a Velero-style approach? Where are files
+stored? Might be time to introduce a KUDO-specific Minio instance.
+
+Filtering is an important part of diagnostics collection. It enables diagnostics
+to be portably sent to third parties that should not have sensitive information
+that logs and files can contain.
+
+By default, KUDO filters all resources (and custom resources) of values
+contained within the KUDO Instance's secrets. This is configurable with the
+`diagnostics.filterSecrets` key.
+
+There may be other fields that need to be filtered. To solve for this, KUDO
+introduces the `diagnostics.bundle.filters` key in `operator.yaml`, which
+contains a list of filters that files pass through before writing to disk.
+Custom filters use either a regular expression or an object reference and
+JSONPath to derive values to filter.
+
+All filtered values appear as `**FILTERED**` in relevant logs and files.
+
+### More Notes
+
+- Do we need to introduce a notion of the collector or controller manager
+  signing and/or encrypting bundles? TBD.
+
+#### Preflight Checks
+
+## Resources
+
+### bundle.resources
+
+An individual bundle resource is represented as a list inside of the
+`diagnostics.bundle.resources` key. Resources ALWAYS have the following keys:
+
+- **name**: The human-readable name of the file.
+- **key**: The machine-readable name of the file. This is used for both
+  references (if needed in the future) and filenames. Extension is OPTIONAL,
+  but may be useful for inferring mime types.
+- **kind**: The kind of bundle item.
+- **spec**: The attributes of a particular kind. This is different for every
+  kind.
+
+Also, specs may include an `objectRef`. It ALWAYS has the following keys:
+
+- **kind**: The Kubernetes Kind referenced. For example, this may be a
+  Deployment, Pod, StatefulSet, or other resource.
+- **name**: The name of the object. This is a templated field, and has the same
+  template environment as operator templates.
+
+### bundle.resources.Copy
+
+- **path**: Absolute path inside of the referenced pods.
+- **objectRef**
+
+### bundle.resources.Command
+
+- **command**: Command to run. May be a string or an array.
+- **objectRef**
+
+### bundle.resources.Task
+
+- **taskRef**: Name of the task to run. **NOTE**: We MAY need a Pause and Resume
+  task to be able to copy files and run commands during the running of a task.
+  Otherwise, we may want to make this an arbitrary job.
+
+### bundle.resources.HTTP
+
+- **serviceRef**: Object containing references to a Kubernetes service. This is
+  scoped to KUDO-only services.
```

Comment:
Please clarify what a "KUDO-only service" is. Personally I'm puzzled by this statement.

Reply:
I am puzzled as well. @gerred could you please shed some light here?

Reply:
We're just assuming that the only Kubernetes services we can reached were made _by_ KUDO.